lib/node-token-validation/src/api/TokenValidator.ts

```diff
@@ -0,0 +1,138 @@
+
+export class TokenValidator {
+    private config: TokenValidationConfiguration;
+    protected logger: Logger;
+    protected networkInterface: INetworkModule;
+    protected openIdConfigProvider: OpenIdConfigProvider;
+
+    constructor(configuration: Configuration) {
+        this.config = buildConfiguration(configuration);
+        this.logger = new Logger(this.config.system.loggerOptions, name, version);
+        this.networkInterface = this.config.system.networkClient,
+        this.openIdConfigProvider = new OpenIdConfigProvider(this.config.auth.authority, this.networkInterface, this.logger);
+    }
+
+    async validateToken(token: string, options: TokenValidationParameters): Promise<TokenValidationResponse> {
+        this.logger.verbose("validateToken called");
+        
+        if (!token) {
+            throw ValidationConfigurationError.createMissingTokenError();
+        }
+
+        const validationParams: ValidationParameters = await buildTokenValidationParameters(options, this.config);
+        this.logger.verbose("ValidationParams built");
+            
+        const jwks = await this.getJWKS(validationParams);
+
+        const jwtVerifyParams: JWTVerifyOptions = {
+            algorithms: validationParams.validAlgorithms,
+            issuer: validationParams.validIssuers,
+            audience: validationParams.validAudiences,
+            subject: validationParams.subject
+            // Figure out types here
+        };
+
+        const { payload, protectedHeader } = await jwtVerify(token, jwks, jwtVerifyParams);
+
+        this.validateClaims(payload, validationParams);
+
+        return {
+            protectedHeader,
+            payload
+        };
+    }
+    
+    async getJWKS(validationParams: ValidationParameters): Promise<any> {
+        this.logger.verbose("getJWKS called");
+        
+        // Prioritize keystore or jwksUri if provided
+        if (validationParams.issuerSigningKeys) {
+            this.logger.verbose("JweKeyStore provided");
+            return validationParams.issuerSigningKeys;
+        } 
+        
+        if (validationParams.issuerSigningJwksUri) {
+            this.logger.verbose("JwksUri provided");
+            return createRemoteJWKSet(new URL(validationParams.issuerSigningJwksUri));
+        }
+
+        // Do resiliency well-known endpoint thing here
+        const retrievedJwksUri: string = await this.openIdConfigProvider.fetchJwksUriFromEndpoint();
+        this.logger.verbose("Creating Jwks from default");
+        return createRemoteJWKSet(new URL(retrievedJwksUri));
+    
+    }
+ 
+    async validateClaims(payload: JWTPayload, validationParams: ValidationParameters): Promise<void> {
+        this.logger.verbose("validateClaims called");
+
+        // Validate nonce
+        if (payload.nonce) {
+            if (!validationParams.nonce) {
+                this.logger.verbose("Nonce detected on token, but nonce not set in validationParams");
```

Comment:
To be clear, this is not an error condition?

Reply:
Upon further examination, this would be better as an error condition. Thanks for the catch!